ONSdigital/census-mock-case-api-service
# CR-1089 amend addData method so that cases can be replaced

# Motivation and Context

This change to the mock case service is required in order to mimic RM in the following scenario:

Given the case does not exist in the cache
And the case exists in RM
When the case address details are modified
And the case modified event is sent to RM and RM <does immediately> action it
When the call is made to fetch the case again from GetCaseByUPRN
Then GetCaseByUPRN gets the modified case from RM

NB. This scenario will be added to the cc cuc repo.

# What has changed

The following POST endpoint (previously the add endpoint) has been amended so that cases can be either added or replaced (if they are an existing case). It is works locally with the following URL:

http://localhost:8161/cases/data/cases/save

It still requires a request body, which is of type List<CaseContainerDTO>, such as this:

[{
        "caseRef": "123123002",
        "arid": "2344266233",
        "estabArid": "AABBCC",
        "estabType": "ET",
        "uprn": "1347459999",
        "createdDateTime": "2019-04-14T13:45:26.564+01:00",
        "addressLine1": "Happy House Three",
        "addressLine2": "89 Harbour Street",
        "addressLine3": "Somewhere",
        "townName": "Portsmouth",
        "postcode": "G1 2AA",
        "organisationName": "ON",
        "addressLevel": "E",
        "abpCode": "AACC",
        "latitude": "41.40338",
        "longitude": "2.17403",
        "oa": "EE22",
        "lsoa": "x1",
        "msoa": "x2",
        "lad": "H1",
        "caseEvents": [
            {
                "id": "101",
                "eventType": "CASE_UPDATED",
                "description": "Initial creation of case",
                "createdDateTime": "2019-04-14T13:45:26.564+01:00"
            },
            {
                "id": "102",
                "eventType": null,
                "description": "Create Household Visit",
                "createdDateTime": "2019-04-14T13:45:26.564+01:00"
            }
        ],
        "id": "3305e937-6fb1-4ce1-9d4c-077f147789ac",
        "caseType": "HH",
        "addressType": "SPG",
        "region": "E",
        "collectionExerciseId": "49871667-117d-4a63-9101-f6a0660f73f6",
        "surveyType": "CENSUS",
        "handDelivery": false
    }]




# How to test?

Test the changes by firstly running one of the GET case endpoints in Postman e.g.

http://localhost:8161/cases/3305e937-6fb1-4ce1-9d4c-077f147789ac

Then run the save endpoint in Postman, as described above, and then re-run the GET endpoint to check that the case has been modified.
